### PR TITLE
net: openthread: Adding `diag transmit` command.

### PR DESCRIPTION
**Pull request by canisLupus1313**
Commit adds `diag transmit` used for transmission IEEE802154 packets in specific amount and interval.

### CODE REVIEW

File: modules/openthread/platform/diag.c [hunk 1]
```diff
@@ -313,3 +339,77 @@ static otError startModCarrier(otInstance *aInstance, uint8_t aArgsLength, char
...
+		sTransmitMode = kDiagTransmitModeIdle;
+		otPlatRadioReceive(aInstance, sChannel);
+	} else if (strcmp(aArgs[0], "start") == 0) {
+		if (sTransmitMode != kDiagTransmitModeIdle) {
+			error = OT_ERROR_INVALID_STATE;
+		}
+
+		otPlatAlarmMilliStop(aInstance);
+		sTransmitMode = kDiagTransmitModePackets;
+		sTxCount = sTxRequestedCount;
```
Comment (maciejbaczmanski) [inline modules/openthread/platform/diag.c:368]: is `sTxCount` used? I can see it being assigned only here

**Review by maciejbaczmanski — COMMENTED** [review]
I think it might be worth exiting early in most places where we assign error value (like `OT_ERROR_INVALID_ARGS` etc. For example now we allow setting interval to a negative value and will cast it to `uint32_t`

File: modules/openthread/platform/diag.c [hunk 2]
```diff
@@ -8,20 +8,41 @@
...
 
 #include <openthread/error.h>
+#include <openthread/platform/alarm-milli.h>
 #include <openthread/platform/diag.h>
+#include <openthread/platform/radio.h>
 
 #include "platform-zephyr.h"
 #include "zephyr/sys/util.h"
 
+typedef enum {
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/diag.c]: droppd the typedef and changed the name convention

File: modules/openthread/platform/diag.c [hunk 3]
```diff
@@ -8,20 +8,41 @@
...
+static DiagTrasmitMode sTransmitMode = kDiagTransmitModeIdle;
+static uint8_t sChannel = 20;
+static uint32_t sTxPeriod = 1;
+static int32_t sTxCount = 0;
+static int32_t sTxRequestedCount = 1;
 
 static otError startModCarrier(otInstance *aInstance, uint8_t aArgsLength, char *aArgs[]);
+static otError processTransmit(otInstance *aInstance, uint8_t aArgsLength, char *aArgs[]);
+
+static otError parseLong(char *aArgs, long *aValue)
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/diag.c]: fixed.

File: modules/openthread/platform/diag.c [hunk 4]
```diff
@@ -57,6 +78,10 @@ otError otPlatDiagProcess(otInstance *aInstance, uint8_t aArgsLength, char *aArg
 	}
 #endif
 
+	if (strcmp(aArgs[0], "transmit") == 0) {
+		return processTransmit(aInstance, aArgsLength - 1, aArgs + 1);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/diag.c:82]: fixed.

File: modules/openthread/platform/diag.c (hunk 1 above)
Comment (canisLupus1313) [reply modules/openthread/platform/diag.c:422]: Pleae see updated implementation.

File: modules/openthread/platform/diag.c [hunk 5]
```diff
@@ -313,3 +339,77 @@ static otError startModCarrier(otInstance *aInstance, uint8_t aArgsLength, char
...
+{
+	otError error = OT_ERROR_NONE;
+
+	if (aArgsLength == 0) {
+		diag_output("transmit will send %" PRId32 " diagnostic messages with %" PRIu32
+			    " ms interval\r\nstatus 0x%02x\r\n",
+			    sTxRequestedCount, sTxPeriod, error);
+	} else if (strcmp(aArgs[0], "stop") == 0) {
+		if (sTransmitMode == kDiagTransmitModeIdle) {
+			error = OT_ERROR_INVALID_STATE;
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/diag.c]: Please tak a look at new state machine.

File: modules/openthread/platform/diag.c [hunk 6]
```diff
@@ -313,3 +339,77 @@ static otError startModCarrier(otInstance *aInstance, uint8_t aArgsLength, char
...
 #endif
+
+static otError processTransmit(otInstance *aInstance, uint8_t aArgsLength, char *aArgs[])
+{
+	otError error = OT_ERROR_NONE;
+
+	if (aArgsLength == 0) {
+		diag_output("transmit will send %" PRId32 " diagnostic messages with %" PRIu32
+			    " ms interval\r\nstatus 0x%02x\r\n",
+			    sTxRequestedCount, sTxPeriod, error);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/diag.c]: Good point fixed.

File: modules/openthread/platform/diag.c [hunk 7]
```diff
@@ -291,25 +319,152 @@ otError otPlatDiagGpioGetMode(uint32_t aGpio, otGpioMode *aMode)
...
 
 	return platformRadioTransmitModulatedCarrier(aInstance, enable, data);
 }
 
 #endif
+
+void otPlatDiagAlarmCallback(otInstance *aInstance)
+{
+	uint32_t now;
+	otRadioFrame *sTxPacket;
```
Comment (maciejbaczmanski) [inline modules/openthread/platform/diag.c:352]:
```suggestion
	otRadioFrame *txPacket;
```
`sTxPacket` suggests that variable is static

File: modules/openthread/platform/diag.c [hunk 8]
```diff
@@ -291,25 +319,152 @@ otError otPlatDiagGpioGetMode(uint32_t aGpio, otGpioMode *aMode)
...
+			    sTxRequestedCount, sTxPeriod);
+
+	} else if (strcmp(aArgs[0], "stop") == 0) {
+		if (sTransmitMode == diag_transmit_mode_idle) {
+			return OT_ERROR_INVALID_STATE;
+		}
+
+		otPlatAlarmMilliStop(aInstance);
+		diag_output("diagnostic message transmission is stopped\r\nstatus 0x%02x\r\n",
+			    error);
```
Comment (maciejbaczmanski) [inline modules/openthread/platform/diag.c:411]:
```suggestion
		diag_output("diagnostic message transmission is stopped\r\n");
```
`error` is not updated in this case so it is always `OT_ERROR_NONE`

File: modules/openthread/platform/diag.c [hunk 9]
```diff
@@ -291,25 +319,152 @@ otError otPlatDiagGpioGetMode(uint32_t aGpio, otGpioMode *aMode)
...
+		}
+
+		otPlatAlarmMilliStop(aInstance);
+		sTransmitMode = diag_transmit_mode_packets;
+		sTxCount = sTxRequestedCount;
+		now = otPlatAlarmMilliGetNow();
+		otPlatAlarmMilliStartAt(aInstance, now, sTxPeriod);
+		diag_output("sending %" PRId32 " diagnostic messages with %" PRIu32
+			    " ms interval\r\nstatus 0x%02x\r\n",
+			    sTxRequestedCount, sTxPeriod, error);
```
Comment (maciejbaczmanski) [inline modules/openthread/platform/diag.c:427]:
```suggestion
		diag_output("sending %" PRId32 " diagnostic messages with %" PRIu32
			    " ms interval\r\n",
			    sTxRequestedCount, sTxPeriod);
```
`error` is not updated in this case so it is always `OT_ERROR_NONE`

File: modules/openthread/platform/diag.c [hunk 10]
```diff
@@ -291,25 +319,152 @@ otError otPlatDiagGpioGetMode(uint32_t aGpio, otGpioMode *aMode)
...
+			return error;
+		}
+
+		if (value <= 0) {
+			return OT_ERROR_INVALID_ARGS;
+		}
+		sTxPeriod = (uint32_t)(value);
+		diag_output("set diagnostic messages interval to %" PRIu32
+			    " ms\r\nstatus 0x%02x\r\n",
+			    sTxPeriod, error);
```
Comment (maciejbaczmanski) [inline modules/openthread/platform/diag.c:445]:
```suggestion
		diag_output("set diagnostic messages interval to %" PRIu32
			    " ms\r\n", sTxPeriod);
```
we get here only if `error==OT_ERROR_NONE`

File: modules/openthread/platform/diag.c [hunk 11]
```diff
@@ -291,25 +319,152 @@ otError otPlatDiagGpioGetMode(uint32_t aGpio, otGpioMode *aMode)
...
+			return error;
+		}
+
+		if ((value <= 0) && (value != -1)) {
+			return OT_ERROR_INVALID_ARGS;
+		}
+
+		sTxRequestedCount = (uint32_t)(value);
+		diag_output("set diagnostic messages count to %" PRId32 "\r\nstatus 0x%02x\r\n",
+			    sTxRequestedCount, error);
```
Comment (maciejbaczmanski) [inline modules/openthread/platform/diag.c:464]:
```suggestion
		diag_output("set diagnostic messages count to %" PRId32 "\r\n",
			    sTxRequestedCount);
```
we get here only if `error==OT_ERROR_NONE`

File: modules/openthread/platform/diag.c [hunk 12]
```diff
@@ -8,20 +8,44 @@
...
 
 #include "platform-zephyr.h"
 #include "zephyr/sys/util.h"
 
+enum {
+	diag_transmit_mode_idle,
+	diag_transmit_mode_packets,
+	diag_transmit_mode_carrier,
+	diag_transmit_mode_modcarrier
+
```
Comment (rlubos) [inline modules/openthread/platform/diag.c:23]: nit: Enum values should be all uppercase.

Comment (canisLupus1313) [reply modules/openthread/platform/diag.c]: fixed

**Review by rlubos — DISMISSED** [review]

**Review by maciejbaczmanski — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**